rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,325 @@
+use provided reference implementation as a source of ground truth (that can run
+on custom hardware out of the box!) when developing custom implementation on
+their hardware.
+
+4) *Reduced implementation dependencies*: Operation maintainer can change the
+implementation without the fear of breaking the downstream tools pattern
+recognition (since there isn’t any.)
+
+5) *Forward compatibility*: Operation maintainer can even add an atomic
+implementation without breaking existing tools. No overhead in keeping the basic
+implementation available.
+
+6) *Simpler automatic conversion* As tensorflow moves forward it is conceivable
+that its computation description language (today: graph def, tomorrow:MLIR) 
+will be fully  be separated from the underlying tensorflow kernel implementations.
+This kind of standard attributes allow for more  automatic conversions
+should the need arise.
+
+7) Does not change op developer workflow and can be introduced incrementally to
+existing code.
+
+### More on forward compatibility
+
+Suppose later, the maintainer of the op decides that certain composite op
+deserves its own atomic op due to its usefulness. If we have standardized
+attributes like the one above, the XLA compiler can check if op with the name
+that this function “implements” is available and substitute it at runtime. The
+significant advantage is that even old serialized models will be able to take
+advantage of improved implementations. As the op gains more adoption and its
+support becomes more widespread the previous composite definition can be
+deprecated or retired. However such an approach allows both implementations to
+coexist for unlimited period of time and the same graph can be both: efficient
+on the platforms that support it, and “not-broken” on the platforms that don’t.
+
+### Alternatives considered
+
+#### Alternative 0: Do nothing
+
+Everyone just continues with adhoc implementations.
+
+##### Pros:
+
+*   No need for RFC or new public API
+
+##### Cons:
+
+*   Downstream tools are stuck between (a) trying to implement ops, even if the
+    efficiency is not yet a concern. In which case this just redos
+    implementation in the downstream tooling using its own language. Or (b)
+    trying to extract certain graph patterns that can be optimizing and
+    following the moving target. Once such pattern is extracted there is a
+    pressure on op-maintainers to freeze his implementation to avoid breaking
+    the patterns used by the downstream tools. Or (c) Tring to roll out their
+    own extension such as `OpHint` for tflite.
+
+*   New ops that could push the industry forward however are stymied due to lack
+    efficient implementation, software vendors are not interested in
+    implementation until the ops become popular enough, creating a vicious
+    cycle.
+
+#### Alternative 1: Recommend each new op to be implemented as a new atomic node in graphdef.
```

Comment:
Not sure I follow this alternative. In many use cases, depending on the front end you are programming in, the generated graph and the atomicity of the nodes in it, might be out of the user's control. Does this only work when operating at low enough level in the system?

Reply:
Right, this particular alternative, is inherently limiting how new ops could be created, and therefore is probably not a good option. It certainly would be out of reach of anyone who isn't fluent in using raw tensorflow. (and comes with a slew of other limitations)